### DISCUSSION TITLE
Storing search-variants?

### DISCUSSION THREAD

**Discussion by dg9vh** [magicbug/Cloudlog#1140] [Ideas / Feature Requests]
Hi,

I often use the extended search for searching different situations with combinations of search criteria. Because of this "regular behavior" every x weeks I think it would be an effort, if we can store such search-variants as templates on individually base to just select from a select-box and click on search to have the same search weeks later.

Would this be an idea and who will accept the challenge? 

73 de Kim
DG9VH

**Comment by AndreasK79** [comment]
Interesting idea Kim. Not sure what the best way for this is. One idea that comes to mind, is to store the finished generated query in the database. Then this can be ran again next time...

Andreas.

**Comment by dg9vh** [comment]
This could be a way... I think, this way could also be used to parse the
sql and rebuild the query in the query designer of
cloudlog-extended-search... but don't know.

Am Mo., 30. Aug. 2021 um 13:58 Uhr schrieb Andreas Kristiansen <
***@***.***>:

> Interesting idea Kim. Not sure what the best way for this is. One idea
> that comes to mind, is to store the finished generated query in the
> database. Then this can be ran again next time...
>
> Andreas.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/magicbug/Cloudlog/discussions/1140#discussioncomment-1253771>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADKN5OS5TWVCRKAQT5SI363T7NW6FANCNFSM5CUKVZSA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>


-- 
Amateurfunk ist das, was uns verbindet! Bleiben wir in Kontakt!
Ortsverband Q03, Neunkirchen

**Comment by magicbug** [comment]
I'm not sure the solution to this but agree would be nice to be able to recall searches, also need to add a QSO count to the search etc.

**Comment by dg9vh** [comment]
QSO-Count could easily be done if the results-table would be converted to a
DataTables-Table... Here you can also sort by clicking on columns...

Maybe it would be ok to have a way to define the search by storing a
corresponding SQL-query-part (for the where-clause)... actually not really
have the time to think this over.

Am Mi., 8. Sept. 2021 um 22:54 Uhr schrieb Peter Goodhall <
***@***.***>:

> I'm not sure the solution to this but agree would be nice to be able to
> recall searches, also need to add a QSO count to the search etc.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/magicbug/Cloudlog/discussions/1140#discussioncomment-1298261>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADKN5OQQXMKCF2UMV5MNN33UA7EPJANCNFSM5CUKVZSA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>


-- 
Amateurfunk ist das, was uns verbindet! Bleiben wir in Kontakt!
Ortsverband Q03, Neunkirchen

**Comment by AndreasK79** [comment]
I'll move the discussion here. I have a very crude version running. That means that saving a query and running a query works.
I'm storing the final compiled query, so edit is not possible. To do that, we need a parser to rebuild the query. That is not something I have in mind to write now.

Still a lot of work left to get this to a usable state for the end-user

**Comment by AndreasK79** [comment]
@dg9vh teaser:
![image](https://user-images.githubusercontent.com/6977712/134678979-248bbfef-46ba-4e6c-bb2a-e8b3e6bb2b25.png)

**Reply by dg9vh** [reply]
Yes, it looks very tasty :-) can't see the "store" button... but I think this will follow :-)

**Reply by AndreasK79** [reply]
It will only show when you have built a query and clicked search:
![image](https://user-images.githubusercontent.com/6977712/134679422-e653abe9-7e57-4f9c-add7-d57893027b2f.png)

**Reply by dg9vh** [reply]
It may be better to switch to a select-box-list-choser instead of buttons... maybe... but I don't think that someone will store more than 10 queries... let's try it with the button-style first :-)

**Reply by dg9vh** [reply]
> It will only show when you have built a query and clicked search:
> ![image](https://user-images.githubusercontent.com/6977712/134679422-e653abe9-7e57-4f9c-add7-d57893027b2f.png)

ok :-)

**Reply by AndreasK79** [reply]
I am open to suggestions, but a dropdown might be better. Specially if the description is a long one.

The datatable result is currently only available from running the stored queries. Still a long way to go.

**Comment by AndreasK79** [comment]
![image](https://user-images.githubusercontent.com/6977712/134761659-bfa2ef0d-49b2-4f6f-91e4-bb45ab5eb020.png)
Dropdown for stored queries, and loading indicators on buttons.

**Comment by dg9vh** [comment]
Looks good

Andreas Kristiansen ***@***.***> schrieb am Sa., 25. Sept.
2021, 08:42:

**Comment by AndreasK79** [comment]
Only description editing is left:
![image](https://user-images.githubusercontent.com/6977712/134778126-3901497d-f1a2-4d1d-a6dd-33e7c5dc0017.png)

**Comment by dg9vh** [comment]
I think this looks very good and editing the sql-query would also be
helpful (for those that are able to do this :-)

Am Sa., 25. Sept. 2021 um 18:10 Uhr schrieb Andreas Kristiansen <
***@***.***>:

> Only description editing is left:
> [image: image]
> <https://user-images.githubusercontent.com/6977712/134778126-3901497d-f1a2-4d1d-a6dd-33e7c5dc0017.png>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/magicbug/Cloudlog/discussions/1140#discussioncomment-1383855>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADKN5OXZHVXIPUGUKLPU3P3UDXX5JANCNFSM5CUKVZSA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>


-- 
Amateurfunk ist das, was uns verbindet! Bleiben wir in Kontakt!
Ortsverband Q03, Neunkirchen

**Reply by AndreasK79** [reply]
Well, I'm debating with myself if we should allow editing of the query or not. 

1. You can put whatever you like in the query. If you know SQL, no problem, but you can delete whatever you want.
2. You can potentially screw up your installation, if you don't know what you are doing.
3. And if someone wants to prank you, they can just edit the query, and all your contacts are gone when you run it.

I know these are edge-cases though. As a safety precaution, I could check for select, but that will not stop a query like select 1=1; delete from station_profile; Then again, I can check that it contains select and not delete.

I need some input here. @dg9vh ? @magicbug ? Anyone else?

**Comment by dg9vh** [comment]
I think for the first step two options for
a) storing a query and
b) deleting a query would last for a while as a first instance.

If we have time and capacity, we could implement a parser to decompile the
stored SQL into a corresponding rule-set for the web-ui.

But having a) and b) would solve the problem for a first shot.

Only my opinion :-)

Kim

Am Sa., 25. Sept. 2021 um 18:58 Uhr schrieb Andreas Kristiansen <
***@***.***>:

> Well, I'm debating with myself if we should allow editing of the query or
> not.
>
>    1. You can put whatever you like in the query. If you know SQL, no
>    problem, but you can delete whatever you want.
>    2. You can potentially screw up your installation, if you don't know
>    what you are doing.
>    3. And if someone wants to prank you, they can just edit the query,
>    and all your contacts are gone when you run it.
>
> I know these are edge-cases though. As a safety precaution, I could check
> for select, but that will not stop a query like select 1=1; delete from
> station_profile; Then again, I can check that it contains select and not
> delete.
>
> I need some input here. @dg9vh <https://github.com/dg9vh> ? @magicbug
> <https://github.com/magicbug> ? Anyone else?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/magicbug/Cloudlog/discussions/1140#discussioncomment-1383993>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADKN5OQDVP2MW5TWINDT5IDUDX5T3ANCNFSM5CUKVZSA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>


-- 
Amateurfunk ist das, was uns verbindet! Bleiben wir in Kontakt!
Ortsverband Q03, Neunkirchen

**Reply by AndreasK79** [reply]
A and B are done. I've pushed the current branch here https://github.com/AndreasK79/Cloudlog/tree/storing_queries if you would like to test it out.

I just forgot, I've not made migration script, so no point in testing just yet. Edit: now added.

**Comment by AndreasK79** [comment]
Edit is now done (only description right now). The code can be found here:  https://github.com/AndreasK79/Cloudlog/tree/storing_queries
I would love if someone could take it for a testrun.

@magicbug I'm not sure if the PR should go to master or station_logbooks. Some code is based on code that is not in station_logbooks, but it does contain a migration script with number 79. I'm not sure if this would cause problems when stations_logbooks will be merged to master?

**Comment by dg9vh** [comment]
Sorry, I don't have a test-system by myself actually and also actually no
time for setting one up...

Am So., 3. Okt. 2021 um 20:11 Uhr schrieb Andreas Kristiansen <
***@***.***>:

> Edit is now done (only description right now). The code can be found here:
> https://github.com/AndreasK79/Cloudlog/tree/storing_queries
> I would love if someone could take it for a testrun.
>
> @magicbug <https://github.com/magicbug> I'm not sure if the PR should go
> to master or station_logbooks. Some code is based on code that is not in
> station_logbooks, but it does contain a migration script with number 79.
> I'm not sure if this would cause problems when stations_logbooks will be
> merged to master?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/magicbug/Cloudlog/discussions/1140#discussioncomment-1421673>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADKN5OX2DHLD2UPZUXQTCMLUFCMGJANCNFSM5CUKVZSA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>


-- 
Amateurfunk ist das, was uns verbindet! Bleiben wir in Kontakt!
Ortsverband Q03, Neunkirchen

**Reply by AndreasK79** [reply]
No problem. You know it's there, so it will be merged, sooner or later.

**Comment by AndreasK79** [comment]
@dg9vh I've merged it into master. Test it out and give me some feedback.

I know about 2 issues:

- When you are storing a query for the first time, the dropdown isn't created, you need to refresh the page.
- When you delete the last query, the dropdown isn't removed.

These are minor issues, but I'll see if I can do something about them.

**Comment by dg9vh** [comment]
Hm... The first issue I have not seen here :-) the second is something
that's not really a big thing. I stored a query and could rerun it...
Editing works on the name (as designed before, because we did not want to
have the query itself edited). So if someone needs to modify the query, he
should rename the old one, create the new one and delete the old one then.

In my eyes, working fine :-)

Am Mi., 6. Okt. 2021 um 12:26 Uhr schrieb Andreas Kristiansen <
***@***.***>:

> @dg9vh <https://github.com/dg9vh> I've merged it into master. Test it out
> and give me some feedback.
>
> I know about 2 issues:
>
>    - When you are storing a query for the first time, the dropdown isn't
>    created, you need to refresh the page.
>    - When you delete the last query, the dropdown isn't removed.
>
> These are minor issues, but I'll see if I can do something about them.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/magicbug/Cloudlog/discussions/1140#discussioncomment-1435215>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADKN5OS6ZC6WKIJ6NJO37J3UFQP5DANCNFSM5CUKVZSA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
>


-- 
Amateurfunk ist das, was uns verbindet! Bleiben wir in Kontakt!
Ortsverband Q03, Neunkirchen
